src/source/custom_source.js

```diff
@@ -0,0 +1,319 @@
+ *
+ *     async loadTile(tile, {signal}) {
+ *         const url = this.tilesUrl
+ *             .replace('{z}', String(tile.z))
+ *             .replace('{x}', String(tile.x))
+ *             .replace('{y}', String(tile.y));
+ *
+ *         const response = await fetch(url, {signal});
+ *         const data = await response.arrayBuffer();
+ *
+ *         const blob = new window.Blob([new Uint8Array(data)], {type: 'image/png'});
+ *         const imageBitmap = await window.createImageBitmap(blob);
+ *
+ *         return imageBitmap;
+ *     }
+ * }
+ *
+ * map.on('load', () => {
+ *     map.addSource('custom-source', new CustomSource());
+ *     map.addLayer({
+ *         id: 'layer',
+ *         type: 'raster',
+ *         source: 'custom-source'
+ *     });
+ * });
+ */
+
+/**
+ * Optional method called when the source has been added to the Map with {@link Map#addSource}.
+ * This gives the source a chance to initialize resources and register event listeners.
+ *
+ * @function
+ * @memberof CustomSourceInterface
+ * @instance
+ * @name onAdd
+ * @param {Map} map The Map this custom source was just added to.
+ */
+
+/**
+ * Optional method called when the source has been removed from the Map with {@link Map#removeSource}.
+ * This gives the source a chance to clean up resources and event listeners.
+ *
+ * @function
+ * @memberof CustomSourceInterface
+ * @instance
+ * @name onRemove
+ * @param {Map} map The Map this custom source was added to.
+ */
+
+/**
+ * Optional method called after the tile is unloaded from the map viewport. This
+ * gives the source a chance to clean up resources and event listeners.
+ *
+ * @function
+ * @memberof CustomSourceInterface
+ * @instance
+ * @name unloadTile
+ * @param {{ z: number, x: number, y: number }} tile Tile name to unload in the XYZ scheme format.
+ */
+
+/**
+ * Optional method called during a render frame to allow a source to prepare and modify a tile texture if needed.
+ *
+ * @function
+ * @memberof CustomSourceInterface
+ * @instance
+ * @name prepareTile
+ * @param {{ z: number, x: number, y: number }} tile Tile name to prepare in the XYZ scheme format.
+ * @returns {TextureImage} The tile image data as an `HTMLImageElement`, `ImageData`, `ImageBitmap` or object with `width`, `height`, and `data`.
+ */
+
+/**
+ * Called when the map starts loading tile for the current animation frame.
+ *
+ * @function
+ * @memberof CustomSourceInterface
+ * @instance
+ * @name loadTile
+ * @param {{ z: number, x: number, y: number }} tile Tile name to load in the XYZ scheme format.
+ * @param {Object} options Options.
+ * @param {AbortSignal} options.signal A signal object that allows the map to cancel tile loading request.
+ * @returns {Promise<TextureImage>} The tile image data as an `HTMLImageElement`, `ImageData`, `ImageBitmap` or object with `width`, `height`, and `data`.
+ */
+export type CustomSourceInterface<T> = {
+    id: string;
+    type: 'custom',
+    dataType: ?DataType,
+    minzoom: number,
```

Comment:
are any of `minzoom`, `maxzoom`, `scheme`, `tileSize` optional?

Reply:
Yep, they are optional, good catch